feat: Alternative representations for nullable protobuf fields

lucasbru opened the pull request:
### Description 

The protobuf support in schema registry by default represents an optional field (corresponding in KSQL to a nullable SQL column) by a singular protobuf field. For primitive types, singular protobuf fields cannot distinguish default values (such as `0` for `int32` and `""` for `string`) from `null`.

This change introduces two properties for the KSQL `CREATE` statements that can enable two alternative representations of nullable fields. The properties are named `KEY_PROTOBUF_NULLABLE_REPRESENTATION`and `VALUE_PROTOBUF_NULLABLE_REPRESENTATION` which control the nullable behavior for keys and value serialization, respectively. Both new properties can take the values `OPTIONAL`, to use protobuf's `optional` keyword, or `WRAPPER`, to use nullable protobuf wrappers (e.g. using `google.protobuf.StringValue` for strings). Setting the properties will allow KSQL and downstream consumers to clearly distinguish between default values and `null`. Leaving the properties undefined will let KSQL use singular protobuf fields - the current default behavior.

Fixes #9333.

### Testing done 

New unit tests are added to cover that the properties are propagated from the `CREATE` statement to the `Serde` implementation. We use query translation tests to test end-to-end test cases.

### Reviewer checklist

- [x] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [x] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

vcrfxia commented on ksqldb-parser/src/main/java/io/confluent/ksql/parser/properties/with/CreateSourceAsProperties.java:
```diff
@@ -154,6 +155,22 @@ public Map<String, String> getKeyFormatProperties(
       builder.put(ProtobufProperties.UNWRAP_PRIMITIVES, ProtobufProperties.UNWRAP);
     }
 
+    final String nullableRep = props.getString(
```
This property should only be set if the format is PROTOBUF, right? Looks like this belongs inside the if-block on line 154.

I'm surprised the existing code doesn't have validation to throw if unknown properties are passed (e.g., if this protobuf-specific property is passed to JSON format). Probably out of scope to add this more generally, though we can consider adding validation to reject statements with `PROTOBUF_NULLABLE_REPRESENTATION` set for non-PROTOBUF formats upfront.

lucasbru replied:
Done

vcrfxia commented on ksqldb-parser/src/main/java/io/confluent/ksql/parser/properties/with/CreateSourceAsProperties.java:
```diff
@@ -154,6 +155,22 @@ public Map<String, String> getKeyFormatProperties(
+          builder.put(ProtobufProperties.NULLABLE_REPRESENTATION,
+              ProtobufProperties.NULLABLE_AS_OPTIONAL);
+          break;
+        default:
```
Throw an exception here, to indicate that this should never happen?

lucasbru replied:
Done

vcrfxia commented on ksqldb-parser/src/main/java/io/confluent/ksql/parser/properties/with/CreateSourceAsProperties.java:
```diff
@@ -180,6 +197,22 @@ public Map<String, String> getValueFormatProperties(final String valueFormat) {
       builder.put(ProtobufProperties.UNWRAP_PRIMITIVES, ProtobufProperties.UNWRAP);
     }
 
+    final String nullableRep = props.getString(
```
Same comments as above, and also for CreateSourceProperties below.

lucasbru replied:
Done

vcrfxia commented on ksqldb-functional-tests/src/main/java/io/confluent/ksql/test/serde/protobuf/ValueSpecProtobufSerdeSupplier.java:
```diff
@@ -15,24 +15,50 @@
     return schemaTranslator.toConnectSchema(schema);
   }
+
+  protected void configureConverter(final Converter c, final boolean isKey) {
```
nit: mark this method as `@Override`, to avoid confusion about where it's used.

lucasbru replied:
Done

vcrfxia commented on ksqldb-functional-tests/src/test/resources/query-validation-tests/protobuf.json:
```diff
@@ -79,6 +79,51 @@
+      ]
+    },
+    {
+      "name": "protobuf primitives with null and nullable as optional",
```
I think QTT has something set up for validating that the schemas produced by the statements are as expected. That could be insightful to add to these tests, since otherwise the behavior of this test and the one below appear identical.

It's not the biggest deal if it's not added, though, since the historic plan spec files (once generated) should include schema validation. Adding it would be a way to make it more explicit that this is the intention behind the test, though.

lucasbru replied:
Done

vcrfxia reviewed (commented):
Thanks @lucasbru ! LGTM with some suggestions inline. Also the suggestion we talked about for making this change more user-friendly by introducing a server-level config with a default setting for this new config, so users don't have to specify the same config again on each new query. (Could be a follow-up PR if you prefer.)

Also, can you update the PR description to mention the name of the new configs, and their acceptable values? The current description does a great job explaining the motivation but doesn't answer "how do I actually use this new feature" at a glance. 

We'll also need docs updates for this feature, ideally in this same PR, so we don't have to worry about which branches the docs changes get merged to.

vcrfxia reviewed (approved):
Thanks @lucasbru -- code changes LGTM! As mentioned previously it'd be great to get docs changes in this same PR if possible (and also a minor update to the PR description).

lucasbru commented:
Thanks @vcrfxia for the super-quick and detailed review! I added the docs and updated the PR description.

Do you usually add Jim Galasyn to the review for a docs change?
Also, I just noticed, I suppose we also need to do this for 'PROTOBUF_NOSR' ?

JimGalasyn reviewed (approved):
LGTM!

vcrfxia commented:
Thanks @lucasbru . Can you also update the docs page [here](https://docs.ksqldb.io/en/latest/developer-guide/ksqldb-reference/create-stream/#stream-properties), and similarly for CREATE TABLE, CREATE STREAM AS SELECT, and CREATE TABLE AS SELECT as well? Feel free to de-dup with the content you added to `serialization.md` however you see fit (e.g., link from `serialization.md` to the other pages if you think that's better than repeating the content).

> I suppose we also need to do this for 'PROTOBUF_NOSR' ?

Hmm, great question. I'm not familiar enough with the `PROTOBUF_NOSR` implementation to say off the top of my head but if the same limitation is present there too, then fixing them together (potentially as a follow-up PR) would be fantastic.

vcrfxia commented on docs/developer-guide/ksqldb-reference/create-stream-as-select.md:
```diff
@@ -189,6 +189,15 @@ table.
+this distinction, set `KEY_PROTOBUF_NULLABLE_REPRESENTATION` to either `OPTIONAL` or `WRAPPER`.
+The schema will be used to serialize keys for the stream created by this `CREATE` statement.
+For more details, see the corresponding section in the
+[Serialization Formats](/reference/serialization#protobuf) documentation.
```
Is this the correct link? It 404s for me on the branch.

vcrfxia reviewed (approved):
> Yes, I can check and add the feature for PROTOBUF_NOSR as well as a follow PR. As for this PR, I will merge it if the CI passes, unless somebody disagrees.

Sounds good. Thanks for the updates!

lucasbru commented:
Failing test is unrelated

vcrfxia commented:
Hey @lucasbru , even though the test that failed on the latest PR build is unrelated to your changes, your PR appears to have broken other tests. In particular, PlannedTestsUpToDateTest. You can see the failure from the commit on master that merged your PR in. 

Also, we typically squash merge for this repo (I'm actually not sure why non-squash merge options are allowed at the moment. We used to only allow squash merges) in order to keep the merge history cleaner. For future reference :)
